### DISCUSSION TITLE
240 V Split Phase Monitoring

### DISCUSSION THREAD

**Discussion by tswany11** [emporia-vue-local/esphome#55] [General]
If I wanted to monitor the total energy of a double pole circuit (240 V in America), I would have to put a CT clamp on both and add the power together, correct? Looking at monitoring the AC power usage and some other larger appliances.

**Comment by flaviut** [comment]
Yes, that would be the most accurate way to do this measurement. However, in many cases, the current only flows phase 1 -> phase 2 (rather than phase 1 -> neutral), in which case you can only put a clamp on one pole and set a multiplier instead.

This is also addressed [in the official installation guide](https://www.emporiaenergy.com/installation-guides), Step 9 (continued).

**Reply by tswany11** [reply]
Thanks for the reply, that makes sense.

**Comment by frdfsnlght** [comment]
I'm going to revive this because there's another way to measure this: use one clamp, but run each of the legs through it, but in opposite directions. This will yield the total current between the legs and take into account appliances that use only one leg for electronics (for example).

I assume in this configuration I wouldn't use a multiplier of 2 on the circuit, but leave it at 1. But what happens with the voltages and power calculations? There doesn't seem to be a way to configure the clamp to indicate it's effectively on 2 phases. Does this mess with any of the reported measurements?

**Reply by brandon-fryslie** [reply]
I'm having difficulty imagining how one would achieve this in a practical sense.  Especially when the mains may be hot.

**Reply by supermarkert** [reply]
Just clip it around the load wires. See article here:
https://docs.iotawatt.com/en/02_06_03/CTbasics.html#v-circuits

**Reply by nanderson97651** [reply]
Following up on this, does the configuration work well?  Or would I be better off just using two clamps for my load?

**Reply by supermarkert** [reply]
Works perfectly for me. If you need pinpoint accuracy of the power used then you should use 2 clamps, because in order to calculate total power you need a Voltage and Current measurement. In the config above using only 1 clamp, you need to pick one of your V phases to use in the calc, but your current sensor is on both phases. So, if your voltages between phases are off by a significant amount, then your power calculations may be off. 
In my case, I don't care about this, and my Voltages on each phase are within 1 V of each other. I'm not using this for billing purposes, but just to (a) know when my clothes dryer is running, and (b) get an approximation of how much energy it uses a month. So more than good enough.
I've considered even averaging the 2 voltage measurements and using that in the calc, but never got around to it.

**Reply by nanderson97651** [reply]
Glad to hear that.  I think I'll take that route for my range.  I believe it will be imbalanced load due to some of the electronics, light, and fan running off of only one of the phases.  I'd rather not use two clamps for it and I can definitely tolerate a ~1% error caused by voltage variation.  I'll consider doing the averaging if I find a dramatic difference between the measured phases.

**Comment by sigurasg** [comment]
A friend of mine has a significant voltage difference reported between the two phases. Using a constant multiplier of 2 will lead to either under or overreporting the power usage. Assuming the same current flows in both phases, the actual power is `P=I*(U1+U2)`, whereas the estimate with the multiplier is `P=2*I*U1`, which is obviously only right when `U1===U2`.
However, it appears that the uC reports the measured power against all three phases in the SensorReading struct, so it should be possible to provide a self-correcting power measurement by summing the estimated power for two phases in a split phase config?

**Reply by sigurasg** [reply]
I looked at the code a little bit, and while I don't know much about ESPHome and conventions there, I think it might be reasonable to model this as a new phase. This might be named RED_BLACK, or e.g. RED_BLACK_SPLIT_PHASE?
In North American split phase configurations you could technically count three phases, P1=L1, P2=L2 and P3=L1-L2, though P3 would be in phase with P1.

**Reply by flaviut** [reply]
what is the cause of the difference between voltages? this could be a symptom of a dangerous situation (floating neutral, what's the voltage between neutral & ground?) or equipment damage at the transformer. You should make sure you have a definitive explanation at the very least.

once that is figured out, it is unlikely that a code change is the right thing to do. I don't have this fully thought out, but you are able to set different phases on different current clamps, even if you'd like to use them for the same load. I think then you can then create a template sensor and use that for the total power.

**Reply by eigenphase** [reply]
I am doing this P=2*I*U1 approach right now but I agree this only works when U1==-U2 (actually _not_ U1==U2).

Now I understand what U2 should be roughly U1 usually with exactly 180 degree phase shift but imperfections and asymmetric loads can indeed make U1 != -U2 ...

> what is the cause of the difference between voltages? this could be a symptom of a dangerous situation (floating neutral, what's the voltage between neutral & ground?) or equipment damage at the transformer.

... and this is not necessarily due to anything wrong.

> once that is figured out, it is unlikely that a code change is the right thing to do. 

I think a code change would, actually, be the right thing. If not, let me ask something different: Why do we even bother measuring both voltages U1 and U2?

If it's so reasonable to assume U1==-U2 in a split system anyway, we would only only need to measure one voltage waveforms and set U2 = -U1 in software, right?


Back to the original suggestion: Right now, in the firmware the voltage waveform is specified as `phase_id`, right?

`- { phase_id: phase_a, input: "16", power: { name: "Circuit 16 Power", id: cir16, internal: true } }`

And these in turn, are defined as:

```
  - platform: emporia_vue
    i2c_id: i2c_a
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK  # Vue device wire color
        calibration: 0.022941687956779
        voltage:
          name: "Phase A Voltage"
          filters: [*moving_avg]
        frequency:
          name: "Phase A Frequency"
          filters: [*moving_avg]
```

Could we do something like this instead?

```
  - platform: emporia_vue
    i2c_id: i2c_a
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK  # Vue device wire color
        calibration: 0.022941687956779
        voltage:
          name: "Phase A Voltage"
          filters: [*moving_avg]
        frequency:
          name: "Phase A Frequency"
          filters: [*moving_avg]
  - id: phase_b  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: RED  # Vue device wire color
        calibration: 0.021873976356348
        voltage:
          name: "Phase B Voltage"
          filters: [*moving_avg]
        phase_angle:
          name: "Phase B Phase Angle"
          filters: [*moving_avg]
  - id: phase_ab
       type: template
       voltage:
          lambda: return id(phase_b).state - id(phase_a).state;

[...]

 - { phase_id: phase_ab, input: "16", power: { name: "Circuit 16 Power", id: cir16, internal: true } }`
```

**Comment by sigurasg** [comment]
> what is the cause of the difference between voltages? this could be a symptom of a dangerous situation (floating neutral, what's the voltage between neutral & ground?) or equipment damage at the transformer. You should make sure you have a definitive explanation at the very least.

Yeah, I've told him as much. He's seeing 115/124, which is a little less than 2% each way from the mean. Is it possible that this is simply a calibration error? 

It's a pretty unusual setup, I think, as this is in a chalet in the hills. To save on copper on the long run from the property boundary to the chalet, there's a step-up transformer (to 660V, I believe) at the entrance of the property, then a split phase step-down transformer at the chalet. All the big loads are 240V, so this is unlikely due to uneven loading.

> once that is figured out, it is unlikely that a code change is the right thing to do. I don't have this fully thought out, but you are able to set different phases on different current clamps, even if you'd like to use them for the same load. I think then you can then create a template sensor and use that for the total power.

Ah - I guess the quick and easy poor man's implementation is to simply define two sensors with the same CT and different phases. This at least would allow him to see whether the energy on the circuits tallies up with the energy on the mains. Right now he's seeing more energy leaving the circuits than enters on the mains :).

That being said, 240V split phase is the norm here in North America, and perhaps it makes sense to cater to it with something less kludgy than a fixed multiplication factor?

**Comment by supermarkert** [comment]
So I had this exact scenario and started looking for a solution. I'm in the US with the standard split phase install, but I have a Whirlpool clothes dryer and the load is different for each phase. The electronics and tumble motor are 120v and use phase 1->N, while the heating element is 240v. So monitoring only one phase (and doubling it) would either be too low or too high.  
Then I stumbled on [this great write-up on the Iotawatt website,](https://docs.iotawatt.com/en/02_06_03/CTbasics.html#v-circuits) and they describe it in particular detail. Including, a nifty solution with a picture at the bottom of the page. This worked perfectly for me. Problem solved.  

tl;dr: use one CT, but loop the second phase through in the opposite direction.

**Reply by mariuszchrobak** [reply]
That two wires in one CT is essentially x2, but with current, not voltage, no? SW is using only one voltage anyway, phase a or phase b (or c)

**Reply by supermarkert** [reply]
Current only. Voltage is measured with the Red/Black wires directly out of the breaker. You match the voltage to the CT in the YAML config.

**Reply by mariuszchrobak** [reply]
That's my point. This is still not a 100% accurate solution. You will get both currents, but power will be off if phases voltage differ.

**Reply by supermarkert** [reply]
Power will not be off (for me) by any significant amount worth caring about. My voltage readings are within 1 V of each other.  If I really wanted to I could probably figure how to average them and use that.  
The only true solution would be to monitor each phase individually, and that requires 2 CTs. Now I can use that one somewhere else.

**Comment by OleSam85** [comment]
Some here on split phase power with 240V devices looking to only use 1 CT (mostly those of us in the US) have proposed averaging the voltages for the power calculation. I think I managed to successfully accomplish this using a lambda filter in my YAML file for my Vue 3.

First I had to add an id to the voltage values in the phases setup so that I could call these later. In an ideal world they probably would be the unfiltered values but for what I'm doing it seems good enough. This is what my _phases_ config looks like now:
```
    phases:
      - id: phase_a  # Verify that this specific phase/leg is connected to correct input wire color on device listed below
        input: BLACK # Vue device wire color
        calibration: 0.01925 # a starting point, may need adjusted to ensure accuracy!
        # To calculate new calibration value use the formula <in-use calibration value> * <accurate voltage> / <reporting voltage>
        voltage:
          name: "${leg_1} Voltage"
          filters: [*throttle_avg, *pos]
          id: phase_a_v
        frequency:
          name: "${leg_1} Frequency"
          filters: [*throttle_avg, *pos]
      - id: phase_b  # see notes above
        input: RED
        calibration: 0.01925
        voltage:
          name: "${leg_2} Voltage"
          filters: [*throttle_avg, *pos]
          id: phase_b_v
        phase_angle:
          name: "${leg_2} Phase Angle"
          filters: [*throttle_avg, *pos]
```
Next in my _ctclamps_ config I created a lambda filter that doubles the value and then multiplies it by a correction factor that adjusts to the power corresponding to the average of the phase voltages. After simplifying the equation it ends up like this when the phase_id is set to phase_a (if the phase_id set to phase_b just swap the locations of the phase voltage ids in the lambda filter):

`      - { phase_id: phase_a, input:  "1", power: { id:  cir1, filters: [ *neg, lambda: return x * (1 + (id(phase_b_v).state / id(phase_a_v).state)); ] } }`

It seems to be working well, though it does seem to make the power signal a bit noisier in HA. This could be because my phase_b voltage is actually noisier, or possibly because the device is calculating power from a faster updating phase_a voltage, and then using filtered signals in the correction factor. 

I'm also not sure if using a filtered signal as part of my lambda filter is effectively slowing down the internal power and energy calculations to the same 5s throttle_avg rate. If anyone knows ESPHome better and can answer this it'd be helpful, since I'd rather have the faster calculation rate for better energy data.

I still think it would be cleaner if there was a phase_ab option to measure the differential voltage, as proposed by @eigenphase here https://github.com/emporia-vue-local/esphome/discussions/55#discussioncomment-14929513. I don't have the ability to contribute this to the project but for those that do, it seems like the best solution for 240V devices (without 120V components).
